src/tests/functional/plugin/cpu/single_layer_tests/gather_tree.cpp

```diff
@@ -0,0 +1,173 @@
+            inp2 = ngraph::builder::makeDynamicInputLayer(ngPrc, secondaryInputType, inputDynamicShapes[1]);
+            inp3 = ngraph::builder::makeDynamicInputLayer(ngPrc, secondaryInputType, inputDynamicShapes[2]);
+            inp4 = ngraph::builder::makeDynamicInputLayer(ngPrc, secondaryInputType, inputDynamicShapes[3]);
+
+            paramsIn.push_back(std::dynamic_pointer_cast<ngraph::opset1::Parameter>(inp2));
+            paramsIn.push_back(std::dynamic_pointer_cast<ngraph::opset1::Parameter>(inp3));
+            paramsIn.push_back(std::dynamic_pointer_cast<ngraph::opset1::Parameter>(inp4));
+        } else if (ngraph::helpers::InputLayerType::CONSTANT == secondaryInputType) {
+            auto maxBeamIndex = inputShape.second.front().at(2) - 1;
+
+            inp2 = ngraph::builder::makeConstant<float>(ngPrc, inputShape.second.front(), {}, true, maxBeamIndex);
+            inp3 = ngraph::builder::makeConstant<float>(ngPrc, {inputShape.second.front().at(1)}, {}, true, maxBeamIndex);
+            inp4 = ngraph::builder::makeConstant<float>(ngPrc, {}, {}, true, maxBeamIndex);
+        } else {
+            throw std::runtime_error("Unsupported inputType");
+        }
+
+        auto operationResult = std::make_shared<ngraph::opset4::GatherTree>(paramsIn.front(), inp2, inp3, inp4);
+
+        ngraph::ResultVector results{std::make_shared<ngraph::opset4::Result>(operationResult)};
+        function = std::make_shared<ngraph::Function>(results, paramsIn, "GatherTree");
+    }
+
+    void generate_inputs(const std::vector<ngraph::Shape>& targetInputStaticShapes) override {
+        inputs.clear();
+        const auto maxBeamIndex = targetInputStaticShapes.front().at(2) - 1;
+        const auto& funcInputs = function->inputs();
+
+        for (size_t i = 0; i < funcInputs.size(); ++i) {
+            auto tensor =
+                ov::test::utils::create_and_fill_tensor(funcInputs[i].get_element_type(),
+                                                        targetInputStaticShapes[i],
+                                                        maxBeamIndex,
+                                                        (i == 2 || i == 3) ? maxBeamIndex / 2 : 0);
+            inputs.insert({funcInputs[i].get_node_shared_ptr(), tensor});
+        }
+    }
+};
+
+TEST_P(GatherTreeLayerCPUTest, CompareWithRefs) {
+    run();
+}
+
+namespace {
+
+const std::vector<InferenceEngine::Precision> netPrecisions = {
+        InferenceEngine::Precision::FP32,
+        InferenceEngine::Precision::I32
+};
+
+const std::vector<InputShape> inputStaticShapes = {{{}, {{5, 1, 10}}}, {{}, {{1, 1, 10}}},
+                                                   {{}, {{20, 1, 10}}}, {{}, {{20, 20, 10}}}};
+
+const std::vector<InputShape> inputDynamicShapes = {{{-1, 1, -1}, {{7, 1, 10}}}, {{-1, 1, {5, 10}}, {{2, 1, 7}}},
+                                                    {{-1, {1, 5}, 10}, {{20, 1, 10}}}, {{-1, -1, -1}, {{20, 20, 15}}}};
```

Comment:
To test dynamic shape change, could we add more cases in the target part of the `Inputshape` ?
e.g.
```
{{-1, -1, -1}, {{20, 20, 15}, {30, 30, 15}}}
```

Reply:
Actually, it's not possible for this particular operation to specify multiple exact shapes for one dynamic shape this way, the cause is the implementation of the base class `SubgraphBaseTest`

Reply:
I see, the tests constructs two kinds of graph. One is all inputs are `Parameter`. One is only the first input is `Parameter`. The multiple target tests may fail in the second case.  But I still think we should find a way to test the runtime dynamic shape change because it's initial target for dynamic supports.

Reply:
Well, that's another thing that it won't work for `InputLayerType::CONSTANT`, only for `InputLayerType::PARAMETER`. But in this case it's possible to write two separate test case instances, one for each `InputLayer` type, and specify multiple exact shapes for `InputLayerType::PARAMETER`. Here the problem is that one input is scalar, it won't even work for `InputLayerType::PARAMETER`

Reply:
In `SubgraphBaseTest::init_input_shapes` this assert will fail if multiple exact shapes are specified due to the fact that one input is scalar

`if (dynShape.rank() == 0) {
            ASSERT_EQ(targetStaticShapeSize, 1) << "Incorrect number of static shapes for static case";`

Reply:
Maybe two way to get grid of this,  first don't change the shape of scalar input, just change other inputs. Or you can assemble the `targetStaticShape` in your own code rather than calling the `init_input_shapes`.

Reply:
done